rh-hideout/pokeemerald-expansion
# Add `OW_AUTO_SIGNPOST` and associated metatile behaviors

INCORPORATES A FEATURE BRANCH - DO NOT SQUASH!

# Description
![Player interacting with signpost with new metatile behaviors](https://github.com/user-attachments/assets/bd227dc5-85eb-4bec-a62a-fd5110bfd9b3)

* Ported @ghoulslash's [`signpost`](https://www.pokecommunity.com/threads/simple-modifications-directory.416647/page-13#post-10327708) branch to expansion
* Added @ShinyDragonHunter's variant of FRLG's Message Box
* Added `OW_AUTO_SIGNPOST` and matching metatile behaviors `MB_SIGNPOST`, `MB_POKEMON_CENTER_SIGN`, `MB_POKEMART_SIGN`, allowing players to replicate the behavior of signposts in FRLG.

# Usage

## `OW_AUTO_SIGNPOST`
In [`include/config/overworld.h`](https://github.com/PokemonSanFran/pokeemerald-expansion/blob/signpost/include/config/overworld.h), developers must set `OW_AUTO_SIGNPOST` to `TRUE`. This will also add three metatile behaviors to the existing list.

## Metatile Behaviors
No tilesets within vanilla Emerald have been modified. Users will need to manually add the new metatile behaviors to the relevent metatiles.

## Functionality
When `OW_AUTO_SIGNPOST` is enabled and the player walking into a space where they are facing north and facing a metatile with `MB_SIGNPOST` behavior, the following occurs:
* The message box changes to match the sign message box from FRLG
* The script assigned to that metatile automatically runs.
	* If the metatile behavior is `MB_POKEMON_CENTER_SIGN` or `MB_POKEMART_SIGN`, generic scripts for those signs are run. 
* After `WALK_AWAY_SIGNPOST_FRAMES` has elapsed, the player can walk away from the sign without finishing the scripts.

## Sign Message Box
The message box from FRLG cannot directly be ripped, as the message box from FRLG and Emerald are slightly different sizes. This uses a modified version that was created by @ShinyDragonHunter.

| Original | Modified |
|---|---|
| ![Red reading the Route 1 Signpost](https://archives.bulbagarden.net/media/upload/c/c8/Route_1_sign.png) | ![Brendan reading the Petalburg City Signpost](https://i.imgur.com/jk99eMI.png) |

# Testing
## Clean Branch
You can recreate this branch by applying a patch or pulling the repo. From a clean version of expansion's upcoming, you can either:

### Patch
`wget https://files.catbox.moe/a96tdk.patch -O signpost.patch ; git apply signpost.patch ; rm signpost.patch`

### Repo
`git remote add psf-expansion https://github.com/PokemonSanFran/pokeemerald-expansion/ ; git pull psf-expansion signpost`

## Manual Tests
After replicating the branch, to recreate my testing environment, you can either directly download the debug script and config file, or manually create the changes.

### Download
`wget https://files.catbox.moe/2mmvxp.bin -O data/tilesets/primary/general/metatile_attributes.bin && wget https://files.catbox.moe/sd7gwo.h -O include/config/overworld.h`

### Manual Testing
* Open the [config file](https://raw.githubusercontent.com/PokemonSanFran/pokeemerald-expansion/signpost/include/config/overworld.h) and change `OW_AUTO_SIGNPOST` to `TRUE`.
* Change the following metatile behaviors in the Primary Tileset:
	* The signpost to have `MB_SIGNPOST`.
	* The Pokémon Center tiles to have `MB_POKEMON_CENTER_SIGN`
	* The Pokémon Mart tiles to have `MB_POKEMART_SIGN`
* Compile the game
* Start a new save
* Run Cheat Start
* Leave the truck
* Player should try
	* leaving the truck
	* walking up to the sign
	* presses A
	* walking away from and back towards the sign
	* walking away from sign before script finishes
	* walking towards Pokémon Center sign
	* walking away from sign before script finishes
	* walking towards Pokémon Mart sign
	* walking away from sign before script finishes

## Verified Scenarios

All videos attempt to show the player:
* leaving the truck
* walking up to the sign
* presses A
* walking away from and back towards the sign
* walking away from sign before script finishes
* walking towards Pokémon Center sign
* walking away from sign before script finishes
* walking towards Pokémon Mart sign
* walking away from sign before script finishes

https://github.com/user-attachments/assets/2b566e1b-2aa4-46b2-953b-31eb7ecc6635

# **People who collaborated with me in this PR**
This was originally [written by](<https://www.pokecommunity.com/threads/simple-modifications-directory.416647/page-13#post-10327708>) by @ghoulslash. The Emerald variant of FRLG's Message Box was created by @ShinyDragonHunter.

# Discord Contact Info
I am `pkmnsnfrn` on Discord.
